touchforms/backend/persistence.py

```diff
@@ -24,32 +26,132 @@ def restore(sess_id, factory, override_state=None):
         state.update(override_state)
     return factory(**state)
 
-# TODO integrate with real caching framework (django ideally)
-def cache_set(key, value, timeout):
-    with open(cache_path(key), 'w') as f:
-        f.write(json.dumps(value).encode('utf8'))
 
-# TODO integrate with real caching framework (django ideally)
+def cache_set(key, value):
+    if key is None:
+        raise KeyError
+    if settings.USES_POSTGRES:
+        postgres_set(key, value)
+    else:
+        with open(cache_get_file_path(key), 'w') as f:
+            f.write(json.dumps(value).encode('utf8'))
+
+
 def cache_get(key):
+    if key is None:
+        raise KeyError
+    if settings.USES_POSTGRES:
+        try:
+            return postgres_lookup(key)
+        except KeyError:
+            return cache_get_file(key)
+    else:
+        return cache_get_file(key)
+
+
+def postgres_select(cursor, key):
+    sel_sql = replace_table("SELECT * FROM %(kwarg)s WHERE sess_id=?")
+    sel_params = [str(key)]
+    cursor.execute(sel_sql, sel_params)
+
+
+def postgres_delete(cursor, key):
+    postgres_select(cursor, key)
+    if cursor.rowcount is not 0:
+        del_sql = replace_table("DELETE FROM %(kwarg)s WHERE sess_id=?")
+        del_params = [str(key)]
+        cursor.execute(del_sql, del_params)
+
+
+def postgres_lookup(key):
+    return postgres_helper(postgres_lookup_command, key)
+
+
+def postgres_lookup_command(cursor, key):
+
+    postgres_select(cursor, key)
+
+    if cursor.rowcount is 0:
+        raise KeyError
+    value = cursor.fetchone()[1]
+
+    jsonobj = json.loads(value.decode('utf8'))
+    return jsonobj
+
+
+def postgres_update_command(cursor, key, value):
+    upd_sql = replace_table("UPDATE %(kwarg)s SET sess_json = ?  WHERE sess_id = ?")
+    upd_params = [json.dumps(value).encode('utf8'), str(key)]
+    cursor.execute(upd_sql, upd_params)
+
+
+def postgres_insert_command(cursor, key, value):
+    ins_sql = replace_table("INSERT INTO %(kwarg)s (sess_id, sess_json) VALUES (?, ?)")
+    ins_params = [str(key), json.dumps(value).encode('utf8')]
+    cursor.execute(ins_sql, ins_params)
+
+
+def postgres_set(key, value):
+    postgres_helper(postgres_set_command, key, value)
+
+
+def postgres_set_command(cursor, key, value):
+
+    postgres_select(cursor, key)
+
+    if cursor.rowcount > 0:
+        postgres_update_command(cursor, key, value)
+    else:
+        postgres_insert_command(cursor, key, value)
```

Comment:
This will not be atomic because each statement is executed in a separate transaction. Not sure if that matters.